### ISSUE TITLE
Multiple typevar union match bug

### ISSUE THREAD

**Issue by hmc-cs-mdrissi** [microsoft/pyright#3613]
**Describe the bug**
```python
from typing import Dict, TypeVar, Union

K = TypeVar('K')
V = TypeVar('V')

T1 = TypeVar('T1')
T2 = TypeVar('T2')

def invert(data: Dict[K, V]) -> Dict[V, K]:
    ...

def f(x: Union[Dict[str, T1], Dict[str, T2]]) -> Union[Dict[T1, str], Dict[T2, str]]:
    return invert(x) # False positive error here of type mismatch
```

**To Reproduce**
Run pyright 1.1.255. Full error message is,

  .../typing_experiments/typevar_confusion.py:13:19 - error: Argument of type "Dict[str, T1@f] | Dict[str, T2@f]" cannot be assigned to parameter "data" of type "Dict[K@invert, V@invert]" in function "invert"
    Type "Dict[str, T1@f] | Dict[str, T2@f]" cannot be assigned to type "Dict[K@invert, V@invert]"
      TypeVar "_VT@dict" is invariant
        Type "T1@f" cannot be assigned to type "T2@f" (reportGeneralTypeIssues)

**Expected behavior**
There should be no error since both types in union are compatible with input type to invert.

**VS Code extension or command-line**
pyright cli 1.1.255

**Additional context**
Is Union ambiguity with type variable matching not allowed? If so I guess a type ignore here works fine.

**Comment by erictraut** [comment]
The current behavior is correct. You cannot pass a union of different dict types in this case because the type parameters for `dict` are invariant. The constraint solver will not be able to find a solution for `K` and `V` that meets these constraints.

Mypy also emits an error in this case.
